File: mobile/src/sync/pullSync.ts
import { getDatabase } from '../db/database';
import { apiRequest } from '../services/apiClient';
import { DatabaseNotifier } from '../db/notifier';
import { Task, Project, Tag, Note, Account, Category, Transaction, Budget } from '../shared/types/domain';

export class PullSync {
  static async pullAll(): Promise<void> {
    await Promise.all([
      this.pullProjects(),
      this.pullTags(),
      this.pullTasks(),
      this.pullNotes(),
      this.pullAccounts(),
      this.pullCategories(),
      this.pullTransactions(),
      this.pullBudgets(),
    ]);

    const db = getDatabase();
    const now = new Date().toISOString();
    db.runSync(
      `INSERT OR REPLACE INTO sync_metadata (table_name, last_synced_at) VALUES ('all', ?)`,
      [now]
    );

    DatabaseNotifier.notify(['tasks', 'projects', 'tags', 'notes', 'accounts', 'categories', 'transactions', 'budgets']);
  }

  static async pullProjects(): Promise<void> {
    try {
      const res = await apiRequest<{ items: Project[] }>('/v1/projects');
      if (!res?.items) return;

      const db = getDatabase();
      const pendingEntities = new Set(
        db.getAllSync<{ entity_id: string }>(
          `SELECT entity_id FROM sync_queue WHERE entity_type = 'project'`
        ).map((r) => r.entity_id)
      );

      db.withTransactionSync(() => {
        for (const item of res.items) {
          if (pendingEntities.has(item.id)) continue;

          const existing = db.getFirstSync<{ updated_at: string }>(
            `SELECT updated_at FROM projects WHERE id = ?`,
            [item.id]
          );

          if (!existing || new Date(item.updated_at) >= new Date(existing.updated_at)) {
            db.runSync(
              `INSERT OR REPLACE INTO projects (id, name, color, is_archived, created_at, updated_at, is_deleted)
               VALUES (?, ?, ?, ?, ?, ?, 0)`,
              [item.id, item.name, item.color || null, item.is_archived ? 1 : 0, item.created_at, item.updated_at]
            );
          }
        }
      });
    } catch (err) {
      console.warn('Pull sync projects error:', err);
    }
  }

  static async pullTags(): Promise<void> {
    try {
      const res = await apiRequest<{ items: Tag[] }>('/v1/tags');
      if (!res?.items) return;

      const db = getDatabase();
      const pendingEntities = new Set(
        db.getAllSync<{ entity_id: string }>(
          `SELECT entity_id FROM sync_queue WHERE entity_type = 'tag'`
        ).map((r) => r.entity_id)
      );

      db.withTransactionSync(() => {
        for (const item of res.items) {
          if (pendingEntities.has(item.id)) continue;

          const existing = db.getFirstSync<{ updated_at: string }>(
            `SELECT updated_at FROM tags WHERE id = ?`,
            [item.id]
          );

          if (!existing || new Date(item.updated_at) >= new Date(existing.updated_at)) {
            db.runSync(
              `INSERT OR REPLACE INTO tags (id, name, color, created_at, updated_at, is_deleted)
               VALUES (?, ?, ?, ?, ?, 0)`,
              [item.id, item.name, item.color || null, item.created_at, item.updated_at]
            );
          }
        }
      });
    } catch (err) {
      console.warn('Pull sync tags error:', err);
    }
  }

  static async pullTasks(): Promise<void> {
    try {
      const res = await apiRequest<{ items: Task[] }>('/v1/tasks?status=all&limit=1000');
      if (!res?.items) return;

      const db = getDatabase();
      const pendingEntities = new Set(
        db.getAllSync<{ entity_id: string }>(
          `SELECT entity_id FROM sync_queue WHERE entity_type = 'task'`
        ).map((r) => r.entity_id)
      );

      db.withTransactionSync(() => {
        for (const item of res.items) {
          if (pendingEntities.has(item.id)) continue;

          const existing = db.getFirstSync<{ updated_at: string }>(
            `SELECT updated_at FROM tasks WHERE id = ?`,
            [item.id]
          );

          if (!existing || new Date(item.updated_at) >= new Date(existing.updated_at)) {
            db.runSync(
              `INSERT OR REPLACE INTO tasks (id, title, description, status, completed_at, due_date, priority, project_id, created_at, updated_at, is_deleted)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 0)`,
              [
                item.id,
                item.title,
                item.description || null,
                item.status,
                item.completed_at || null,
                item.due_date || null,
                item.priority,
                item.project_id || null,
                item.created_at,
                item.updated_at,
              ]
            );

            if (item.tags) {
              db.runSync(`DELETE FROM task_tags WHERE task_id = ?`, [item.id]);
              for (const tag of item.tags) {
                db.runSync(
                  `INSERT OR REPLACE INTO task_tags (task_id, tag_id, created_at) VALUES (?, ?, ?)`,
                  [item.id, tag.id, new Date().toISOString()]
                );
              }
            }

            if (item.subtasks) {
              db.runSync(`DELETE FROM subtasks WHERE task_id = ?`, [item.id]);
              for (const st of item.subtasks) {
                db.runSync(
                  `INSERT OR REPLACE INTO subtasks (id, task_id, title, completed, position, created_at, updated_at, is_deleted)
                   VALUES (?, ?, ?, ?, ?, ?, ?, 0)`,
                  [st.id, item.id, st.title, st.completed ? 1 : 0, st.position, st.created_at, st.updated_at]
                );
              }
            }
          }
        }
      });
    } catch (err) {
      console.warn('Pull sync tasks error:', err);
    }
  }

  static async pullNotes(): Promise<void> {
    try {
      const res = await apiRequest<{ items: Note[] }>('/v1/notes?limit=1000');
      if (!res?.items) return;

      const db = getDatabase();
      const pendingEntities = new Set(
        db.getAllSync<{ entity_id: string }>(
          `SELECT entity_id FROM sync_queue WHERE entity_type = 'note'`
        ).map((r) => r.entity_id)
      );

      db.withTransactionSync(() => {
        for (const item of res.items) {
          if (pendingEntities.has(item.id)) continue;

          const existing = db.getFirstSync<{ updated_at: string }>(
            `SELECT updated_at FROM notes WHERE id = ?`,
            [item.id]
          );

          if (!existing || new Date(item.updated_at) >= new Date(existing.updated_at)) {
            db.runSync(
              `INSERT OR REPLACE INTO notes (id, title, content_md, created_at, updated_at, is_deleted)
               VALUES (?, ?, ?, ?, ?, 0)`,
              [item.id, item.title, item.content_md || '', item.created_at, item.updated_at]
            );

            if (item.tags) {
              db.runSync(`DELETE FROM note_tags WHERE note_id = ?`, [item.id]);
              for (const tag of item.tags) {
                db.runSync(
                  `INSERT OR REPLACE INTO note_tags (note_id, tag_id, created_at) VALUES (?, ?, ?)`,
                  [item.id, tag.id, new Date().toISOString()]
                );
              }
            }
          }
        }
      });
    } catch (err) {
      console.warn('Pull sync notes error:', err);
    }
  }

  static async pullAccounts(): Promise<void> {
    try {
      const res = await apiRequest<{ items: Account[] }>('/v1/accounts');
      if (!res?.items) return;

      const db = getDatabase();
      const pendingEntities = new Set(
        db.getAllSync<{ entity_id: string }>(
          `SELECT entity_id FROM sync_queue WHERE entity_type = 'account'`
        ).map((r) => r.entity_id)
      );

      db.withTransactionSync(() => {
        for (const item of res.items) {
          if (pendingEntities.has(item.id)) continue;

          const existing = db.getFirstSync<{ updated_at: string }>(
            `SELECT updated_at FROM accounts WHERE id = ?`,
            [item.id]
          );

          if (!existing || new Date(item.updated_at) >= new Date(existing.updated_at)) {
            db.runSync(
              `INSERT OR REPLACE INTO accounts (id, name, account_type, currency, archived_at, created_at, updated_at, is_deleted)
               VALUES (?, ?, ?, 'IDR', ?, ?, ?, 0)`,
              [item.id, item.name, item.account_type, item.archived_at || null, item.created_at, item.updated_at]
            );
          }
        }
      });
    } catch (err) {
      console.warn('Pull sync accounts error:', err);
    }
  }

  static async pullCategories(): Promise<void> {
    try {
      const res = await apiRequest<{ items: Category[] }>('/v1/categories');
      if (!res?.items) return;

      const db = getDatabase();
      const pendingEntities = new Set(
        db.getAllSync<{ entity_id: string }>(
          `SELECT entity_id FROM sync_queue WHERE entity_type = 'category'`
        ).map((r) => r.entity_id)
      );

      db.withTransactionSync(() => {
        for (const item of res.items) {
          if (pendingEntities.has(item.id)) continue;

          const existing = db.getFirstSync<{ updated_at: string }>(
            `SELECT updated_at FROM categories WHERE id = ?`,
            [item.id]
          );

          if (!existing || new Date(item.updated_at) >= new Date(existing.updated_at)) {
            db.runSync(
              `INSERT OR REPLACE INTO categories (id, name, category_type, icon, color, archived_at, created_at, updated_at, is_deleted)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, 0)`,
              [item.id, item.name, item.category_type, item.icon || null, item.color || null, item.archived_at || null, item.created_at, item.updated_at]
            );
          }
        }
      });
    } catch (err) {
      console.warn('Pull sync categories error:', err);
    }
  }

  static async pullTransactions(): Promise<void> {
    try {
      const res = await apiRequest<{ items: Transaction[] }>('/v1/transactions?limit=1000');
      if (!res?.items) return;

      const db = getDatabase();
      const pendingEntities = new Set(
        db.getAllSync<{ entity_id: string }>(
          `SELECT entity_id FROM sync_queue WHERE entity_type = 'transaction'`
        ).map((r) => r.entity_id)
      );

      db.withTransactionSync(() => {
        for (const item of res.items) {
          if (pendingEntities.has(item.id)) continue;

          const existing = db.getFirstSync<{ updated_at: string }>(
            `SELECT updated_at FROM transactions WHERE id = ?`,
            [item.id]
          );

          if (!existing || new Date(item.updated_at) >= new Date(existing.updated_at)) {
            db.runSync(
              `INSERT OR REPLACE INTO transactions (id, transaction_type, amount, transaction_date, account_id, transfer_account_id, category_id, description, created_at, updated_at, is_deleted)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 0)`,
              [
                item.id,
                item.transaction_type,
                item.amount,
                item.transaction_date,
                item.account_id,
                item.transfer_account_id || null,
                item.category_id || null,
                item.description || null,
                item.created_at,
                item.updated_at,
              ]
            );
          }
        }
      });
    } catch (err) {
      console.warn('Pull sync transactions error:', err);
    }
  }

  static async pullBudgets(): Promise<void> {
    try {
      const res = await apiRequest<{ items: Budget[] }>('/v1/budgets');
      if (!res?.items) return;

      const db = getDatabase();
      const pendingEntities = new Set(
        db.getAllSync<{ entity_id: string }>(
          `SELECT entity_id FROM sync_queue WHERE entity_type = 'budget'`
        ).map((r) => r.entity_id)
      );

      db.withTransactionSync(() => {
        for (const item of res.items) {
          if (pendingEntities.has(item.id)) continue;

          const existing = db.getFirstSync<{ updated_at: string }>(
            `SELECT updated_at FROM budgets WHERE id = ?`,
            [item.id]
          );

          if (!existing || new Date(item.updated_at) >= new Date(existing.updated_at)) {
            db.runSync(
              `INSERT OR REPLACE INTO budgets (id, category_id, period_start, period_end, amount_limit, created_at, updated_at, is_deleted)
               VALUES (?, ?, ?, ?, ?, ?, ?, 0)`,
              [item.id, item.category_id, item.period_start, item.period_end, item.amount_limit, item.created_at, item.updated_at]
            );
          }
        }
      });
    } catch (err) {
      console.warn('Pull sync budgets error:', err);
    }
  }
}

File: android/app/src/main/java/com/tasker/android/sync/PullSync.kt
package com.tasker.android.sync

import com.tasker.android.data.local.dao.AccountDao
import com.tasker.android.data.local.dao.BudgetDao
import com.tasker.android.data.local.dao.CategoryDao
import com.tasker.android.data.local.dao.NoteDao
import com.tasker.android.data.local.dao.NoteTagDao
import com.tasker.android.data.local.dao.ProjectDao
import com.tasker.android.data.local.dao.RecurringDao
import com.tasker.android.data.local.dao.SubtaskDao
import com.tasker.android.data.local.dao.SyncMetadataDao
import com.tasker.android.data.local.dao.SyncQueueDao
import com.tasker.android.data.local.dao.TagDao
import com.tasker.android.data.local.dao.TargetDao
import com.tasker.android.data.local.dao.TaskDao
import com.tasker.android.data.local.dao.TaskTagDao
import com.tasker.android.data.local.dao.TransactionDao
import com.tasker.android.data.local.entity.AccountEntity
import com.tasker.android.data.local.entity.BudgetEntity
import com.tasker.android.data.local.entity.CategoryEntity
import com.tasker.android.data.local.entity.NoteEntity
import com.tasker.android.data.local.entity.NoteTagEntity
import com.tasker.android.data.local.entity.ProjectEntity
import com.tasker.android.data.local.entity.RecurringTransactionEntity
import com.tasker.android.data.local.entity.SubtaskEntity
import com.tasker.android.data.local.entity.SyncMetadataEntity
import com.tasker.android.data.local.entity.TagEntity
import com.tasker.android.data.local.entity.TargetEntity

import com.tasker.android.data.local.entity.TaskEntity
import com.tasker.android.data.local.entity.TaskTagEntity
import com.tasker.android.data.local.entity.TransactionEntity
import com.tasker.android.remote.api.MoneyApi
import com.tasker.android.remote.api.NoteApi
import com.tasker.android.remote.api.TaskApi
import java.time.Instant
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class PullSync @Inject constructor(
    private val taskApi: TaskApi,
    private val noteApi: NoteApi,
    private val moneyApi: MoneyApi,
    private val projectDao: ProjectDao,
    private val tagDao: TagDao,
    private val taskDao: TaskDao,
    private val subtaskDao: SubtaskDao,
    private val taskTagDao: TaskTagDao,
    private val noteDao: NoteDao,
    private val noteTagDao: NoteTagDao,
    private val accountDao: AccountDao,
    private val categoryDao: CategoryDao,
    private val transactionDao: TransactionDao,
    private val budgetDao: BudgetDao,
    private val recurringDao: RecurringDao,
    private val targetDao: TargetDao,
    private val syncQueueDao: SyncQueueDao,
    private val syncMetadataDao: SyncMetadataDao,
) {
    suspend fun pullAll() {
        pullProjects()
        pullTags()
        pullTasks()
        pullNotes()
        pullAccounts()
        pullCategories()
        pullTransactions()
        pullBudgets()
        pullRecurring()
        pullTargets()

        val now = Instant.now().toString()

        syncMetadataDao.upsert(
            SyncMetadataEntity(
                tableName = "all",
                lastSyncedAt = now
            )
        )
    }

    suspend fun pullProjects() {
        try {
            val response = taskApi.listProjects()
            val pendingEntities = syncQueueDao.getPending().map { it.entityId }.toSet()

            for (dto in response.items) {
                if (pendingEntities.contains(dto.id)) continue

                val existing = projectDao.getById(dto.id)
                if (existing == null || dto.updatedAt >= existing.updatedAt) {
                    projectDao.upsert(
                        ProjectEntity(
                            id = dto.id, name = dto.name, color = dto.color, isArchived = dto.isArchived, createdAt = dto.createdAt, updatedAt = dto.updatedAt, isDeleted = 0
                        )
                    )
                }
            }
        } catch (e: Exception) { }
    }

    suspend fun pullTags() {
        try {
            val response = taskApi.listTags()
            val pendingEntities = syncQueueDao.getPending().map { it.entityId }.toSet()

            for (dto in response.items) {
                if (pendingEntities.contains(dto.id)) continue

                val existing = tagDao.getById(dto.id)
                if (existing == null || dto.updatedAt >= existing.updatedAt) {
                    tagDao.upsert(
                        TagEntity(
                            id = dto.id, name = dto.name, color = dto.color, createdAt = dto.createdAt, updatedAt = dto.updatedAt, isDeleted = 0
                        )
                    )
                }
            }
        } catch (e: Exception) { }
    }

    suspend fun pullTasks() {
        try {
            val response = taskApi.listTasks(status = "all", limit = 1000)
            val pendingEntities = syncQueueDao.getPending().map { it.entityId }.toSet()

            for (dto in response.items) {
                if (pendingEntities.contains(dto.id)) continue

                val existing = taskDao.getById(dto.id)
                if (existing == null || dto.updatedAt >= existing.updatedAt) {
                    taskDao.upsert(
                        TaskEntity(
                            id = dto.id, title = dto.title, description = dto.description, status = dto.status, completedAt = dto.completedAt, dueDate = dto.dueDate, priority = dto.priority, projectId = dto.projectId, createdAt = dto.createdAt, updatedAt = dto.updatedAt, isDeleted = 0
                        )
                    )
                    if (dto.tags != null) {
                        taskTagDao.deleteForTask(dto.id)
                        dto.tags.forEach { tagDto ->
                            taskTagDao.insertAll(listOf(TaskTagEntity(taskId = dto.id, tagId = tagDto.id, createdAt = dto.createdAt)))
                        }
                    }
                    if (dto.subtasks != null) {
                        subtaskDao.deleteForTask(dto.id)
                        val subtaskEntities = dto.subtasks.map { stDto ->
                            SubtaskEntity(id = stDto.id, taskId = dto.id, title = stDto.title, completed = stDto.completed, position = stDto.position, createdAt = stDto.createdAt, updatedAt = stDto.updatedAt, isDeleted = 0)
                        }
                        subtaskDao.upsertAll(subtaskEntities)
                    }
                }
            }
        } catch (e: Exception) { }
    }

    suspend fun pullNotes() {
        try {
            val response = noteApi.listNotes(limit = 1000)
            val pendingEntities = syncQueueDao.getPending().map { it.entityId }.toSet()

            for (dto in response.items) {
                if (pendingEntities.contains(dto.id)) continue

                val existing = noteDao.getById(dto.id)
                if (existing == null || dto.updatedAt >= existing.updatedAt) {
                    val offsetsStr = dto.reminderOffsets?.joinToString(",") ?: "0"
                    noteDao.upsert(
                        NoteEntity(
                            id = dto.id,
                            title = dto.title,
                            contentMd = dto.contentMd,
                            reminderAt = dto.reminderAt,
                            reminderOffsets = offsetsStr,
                            createdAt = dto.createdAt,
                            updatedAt = dto.updatedAt,
                            isDeleted = 0
                        )
                    )
                    if (dto.tags != null) {
                        noteTagDao.deleteForNote(dto.id)
                        dto.tags.forEach { tagDto ->
                            noteTagDao.insertAll(listOf(NoteTagEntity(noteId = dto.id, tagId = tagDto.id, createdAt = dto.createdAt)))
                        }
                    }
                }
            }
        } catch (e: Exception) { }
    }

    suspend fun pullAccounts() {
        try {
            val response = moneyApi.listAccounts()
            val pendingEntities = syncQueueDao.getPending().map { it.entityId }.toSet()

            for (dto in response.items) {
                if (pendingEntities.contains(dto.id)) continue

                val existing = accountDao.getById(dto.id)
                if (existing == null || dto.updatedAt >= existing.updatedAt) {
                    accountDao.upsert(
                        AccountEntity(
                            id = dto.id, name = dto.name, accountType = dto.accountType, currency = dto.currency, archivedAt = dto.archivedAt, createdAt = dto.createdAt, updatedAt = dto.updatedAt, isDeleted = 0
                        )
                    )
                }
            }
        } catch (e: Exception) { }
    }

    suspend fun pullCategories() {
        try {
            val response = moneyApi.listCategories()
            val pendingEntities = syncQueueDao.getPending().map { it.entityId }.toSet()

            for (dto in response.items) {
                if (pendingEntities.contains(dto.id)) continue

                val existing = categoryDao.getById(dto.id)
                if (existing == null || dto.updatedAt >= existing.updatedAt) {
                    categoryDao.upsert(
                        CategoryEntity(
                            id = dto.id, name = dto.name, categoryType = dto.categoryType, icon = dto.icon, color = dto.color, archivedAt = dto.archivedAt, createdAt = dto.createdAt, updatedAt = dto.updatedAt, isDeleted = 0
                        )
                    )
                }
            }
        } catch (e: Exception) { }
    }

    suspend fun pullTransactions() {
        try {
            val response = moneyApi.listTransactions(limit = 1000)
            val pendingEntities = syncQueueDao.getPending().map { it.entityId }.toSet()

            for (dto in response.items) {
                if (pendingEntities.contains(dto.id)) continue

                val existing = transactionDao.getById(dto.id)
                if (existing == null || dto.updatedAt >= existing.updatedAt) {
                    transactionDao.upsert(
                        TransactionEntity(
                            id = dto.id, transactionType = dto.transactionType, amount = dto.amount.toDoubleOrNull() ?: 0.0, transactionDate = dto.transactionDate, accountId = dto.accountId, transferAccountId = dto.transferAccountId, categoryId = dto.categoryId, description = dto.description, createdAt = dto.createdAt, updatedAt = dto.updatedAt, isDeleted = 0
                        )
                    )
                }
            }
        } catch (e: Exception) { }
    }

    suspend fun pullBudgets() {
        try {
            val response = moneyApi.listBudgets()
            val pendingEntities = syncQueueDao.getPending().map { it.entityId }.toSet()

            for (dto in response.items) {
                if (pendingEntities.contains(dto.id)) continue

                val existing = budgetDao.getById(dto.id)
                if (existing == null || dto.updatedAt >= existing.updatedAt) {
                    budgetDao.upsert(
                        BudgetEntity(
                            id = dto.id, categoryId = dto.categoryId, periodStart = dto.periodStart, periodEnd = dto.periodEnd, amountLimit = dto.amountLimit.toDoubleOrNull() ?: 0.0, createdAt = dto.createdAt, updatedAt = dto.updatedAt, isDeleted = 0
                        )
                    )
                }
            }
        } catch (e: Exception) { }
    }

    suspend fun pullRecurring() {
        try {
            val response = moneyApi.listRecurring()
            val pendingEntities = syncQueueDao.getPending().map { it.entityId }.toSet()

            for (dto in response.items) {
                if (pendingEntities.contains(dto.id)) continue

                val existing = recurringDao.getById(dto.id)
                if (existing == null || dto.updatedAt >= existing.updatedAt) {
                    recurringDao.upsert(
                        RecurringTransactionEntity(
                            id = dto.id, transactionType = dto.transactionType, amount = dto.amount.toDoubleOrNull() ?: 0.0, accountId = dto.accountId, categoryId = dto.categoryId, description = dto.description, cadence = dto.cadence, nextDueDate = dto.nextDueDate, endsOn = dto.endsOn, isActive = dto.isActive, lastProcessedOn = dto.lastProcessedOn, createdAt = dto.createdAt, updatedAt = dto.updatedAt, isDeleted = 0
                        )
                    )
                }
            }
        } catch (e: Exception) { }
    }

    suspend fun pullTargets() {
        try {
            val response = moneyApi.listTargets()
            val pendingEntities = syncQueueDao.getPending().map { it.entityId }.toSet()

            for (dto in response.items) {
                if (pendingEntities.contains(dto.id)) continue

                val existing = targetDao.getById(dto.id)
                if (existing == null || dto.updatedAt >= existing.updatedAt) {
                    targetDao.upsert(
                        TargetEntity(
                            id = dto.id,
                            name = dto.name,
                            targetAmount = dto.targetAmount.toDoubleOrNull() ?: 0.0,
                            currentAmount = dto.currentAmount.toDoubleOrNull() ?: 0.0,
                            targetDate = dto.targetDate,
                            categoryId = dto.categoryId,
                            accountId = dto.accountId,
                            color = dto.color,
                            icon = dto.icon,
                            status = dto.status,
                            notes = dto.notes,
                            createdAt = dto.createdAt,
                            updatedAt = dto.updatedAt,
                            isDeleted = 0,
                        )
                    )
                }
            }
        } catch (e: Exception) { }
    }
}

